spec/requests/custom_wizard/steps_controller_spec.rb

```diff
@@ -27,17 +39,121 @@
     expect(response.status).to eq(200)
+    expect(response.parsed_body['wizard']['start']).to eq("step_2")
+  end
+
+  it "returns an updated wizard when condition passes" do
+    new_template = wizard_template.dup
+    new_template['steps'][1]['condition'] = wizard_field_condition_template['condition']
+    CustomWizard::Template.save(new_template, skip_jobs: true)
 
-    wizard = CustomWizard::Builder.new("super_mega_fun_wizard", user).build
-    expect(wizard.start.id).to eq("step_2")
+    put '/w/super-mega-fun-wizard/steps/step_1.json', params: {
+      fields: {
+        step_1_field_1: "Condition will pass"
+      }
+    }
+    expect(response.status).to eq(200)
+    expect(response.parsed_body['wizard']['start']).to eq("step_2")
+  end
+
+  it "returns an updated wizard when condition doesnt pass" do
+    new_template = wizard_template.dup
+    new_template['steps'][1]['condition'] = wizard_field_condition_template['condition']
+    CustomWizard::Template.save(new_template, skip_jobs: true)
+
+    put '/w/super-mega-fun-wizard/steps/step_1.json', params: {
+      fields: {
+        step_1_field_1: "Condition wont pass"
+      }
+    }
+    expect(response.status).to eq(200)
+    expect(response.parsed_body['wizard']['start']).to eq("step_3")
+  end
+
+  it "runs completion actions if user has completed wizard" do
+    new_template = wizard_template.dup
+
+    ## route_to action
+    new_template['actions'].last['run_after'] = 'wizard_completion'
+    new_template['steps'][1]['condition'] = wizard_field_condition_template['condition']
+    new_template['steps'][2]['condition'] = wizard_field_condition_template['condition']
+    CustomWizard::Template.save(new_template, skip_jobs: true)
+
+    put '/w/super-mega-fun-wizard/steps/step_1.json', params: {
+      fields: {
+        step_1_field_1: "Condition wont pass"
+      }
+    }
+    expect(response.status).to eq(200)
+    expect(response.parsed_body['redirect_on_complete']).to eq("https://google.com")
+  end
+
+  it "saves results of completion actions if user has completed wizard" do
+    new_template = wizard_template.dup
+
+    ## Create group action
+    new_template['actions'].first['run_after'] = 'wizard_completion'
+    new_template['steps'][1]['condition'] = wizard_field_condition_template['condition']
+    CustomWizard::Template.save(new_template, skip_jobs: true)
+
+    put '/w/super-mega-fun-wizard/steps/step_1.json', params: {
+      fields: {
+        step_1_field_1: "My cool group"
+      }
+    }
+    expect(response.status).to eq(200)
+
+    put '/w/super-mega-fun-wizard/steps/step_3.json'
+    expect(response.status).to eq(200)
+
+    wizard_id = response.parsed_body['wizard']['id']
+    wizard = CustomWizard::Wizard.create(wizard_id, user)
+    group_name = wizard.submissions.last['action_9']
+    group = Group.find_by(name: group_name)
+    expect(group.full_name).to eq("My cool group")
+  end
+
+  it "detects the final step correctly" do
```

Comment:
We need to add a testcase to ensure whether a step can act as a final and non final step for different wizard paths.

Reply:
@fzngagan I've updated the controller and spec to add more error handling. After looking closer at the code, I don't think it makes sense to have a single test for that case, as it is a combination of other tests. The logic doesn't work without the force_final attribute. Remember that that has a role to play in both in the initial step serialization and the subsequent step submission. The controller is about the second part, the step submission. I'll look at adding more tests for the step serialization piece tomorrow.